controllers/ocirepository_controller.go

```diff
@@ -0,0 +1,888 @@
+		return sreconcile.ResultEmpty, e
+	}
+
+	// Set the internal revision to the remote digest hex
+	revision := imgDigest.Hex
+
+	// Copy the OCI annotations to the internal artifact metadata
+	manifest, err := img.Manifest()
+	if err != nil {
+		e := serror.NewGeneric(
+			fmt.Errorf("failed to parse artifact manifest: %w", err),
+			sourcev1.OCIOperationFailedReason,
+		)
+		conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+		return sreconcile.ResultEmpty, e
+	}
+
+	m := &sourcev1.Artifact{
+		Revision: revision,
+		Metadata: manifest.Annotations,
+	}
+	m.DeepCopyInto(metadata)
+
+	// Mark observations about the revision on the object
+	defer func() {
+		if !obj.GetArtifact().HasRevision(revision) {
+			message := fmt.Sprintf("new digest '%s' for '%s'", revision, url)
+			conditions.MarkTrue(obj, sourcev1.ArtifactOutdatedCondition, "NewRevision", message)
+			conditions.MarkReconciling(obj, "NewRevision", message)
+		}
+	}()
+
+	// Extract the content of the first artifact layer
+	if !obj.GetArtifact().HasRevision(revision) {
+		layers, err := img.Layers()
+		if err != nil {
+			e := serror.NewGeneric(
+				fmt.Errorf("failed to parse artifact layers: %w", err),
+				sourcev1.OCIOperationFailedReason,
+			)
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			return sreconcile.ResultEmpty, e
+		}
+
+		if len(layers) < 1 {
+			e := serror.NewGeneric(
+				fmt.Errorf("no layers found in artifact"),
+				sourcev1.OCIOperationFailedReason,
+			)
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			return sreconcile.ResultEmpty, e
+		}
+
+		blob, err := layers[0].Compressed()
+		if err != nil {
+			e := serror.NewGeneric(
+				fmt.Errorf("failed to extract the first layer from artifact: %w", err),
+				sourcev1.OCIOperationFailedReason,
+			)
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			return sreconcile.ResultEmpty, e
+		}
+
+		if _, err = untar.Untar(blob, dir); err != nil {
+			e := serror.NewGeneric(
+				fmt.Errorf("failed to untar the first layer from artifact: %w", err),
+				sourcev1.OCIOperationFailedReason,
+			)
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			return sreconcile.ResultEmpty, e
+		}
+	}
+
+	conditions.Delete(obj, sourcev1.FetchFailedCondition)
+	return sreconcile.ResultSuccess, nil
+}
+
+// parseRepositoryURL validates and extracts the repository URL.
+func (r *OCIRepositoryReconciler) parseRepositoryURL(obj *sourcev1.OCIRepository) (string, error) {
+	if !strings.HasPrefix(obj.Spec.URL, sourcev1.OCIRepositoryPrefix) {
+		return "", fmt.Errorf("URL must be in format 'oci://<domain>/<org>/<repo>'")
+	}
+
+	url := strings.TrimPrefix(obj.Spec.URL, sourcev1.OCIRepositoryPrefix)
+	ref, err := name.ParseReference(url)
+	if err != nil {
+		return "", err
+	}
+
+	return ref.Context().Name(), nil
+}
```

Comment:
I was trying what happens if I don't provide a reference but have a tag in the URL, we accept the URL with tag but pick the latest tag from the repository.
For example, spec:
```yaml
spec:
  interval: 1m0s
  provider: generic
  timeout: 60s
  url: oci://ghcr.io/stefanprodan/manifests/podinfo:6.1.5
```
Results in:
```yaml
status:
  artifact:
    checksum: 27fdfd6806b2a1cc132d08f3c66e09ec09e4c5d7407111958599988e60125fe9
    lastUpdateTime: "2022-08-03T19:32:14Z"
    metadata:
      org.opencontainers.image.revision: 6.1.8/b3b00fe35424a45d373bf4c7214178bc36fd7872
      org.opencontainers.image.source: https://github.com/stefanprodan/podinfo.git
...
```
Picks the latest tag.
Could this create confusion for the user? Should we add any extra validation?

Reply:
+1

Reply:
Added the same validation as in IRC. Thanks for spotting this.